### PR TITLE
Implement SecureStore class

### PR DESCRIPTION
**Pull request by davidsaada**
Implement SecureStore class.
@dannybenor @yossi2le @offirko @theamirocohen  Please review.

### CODE REVIEW

File: features/storage/kvstore/securestore/SecureStore.cpp [hunk 1]
```diff
@@ -0,0 +1,867 @@
...
+    }
+
+    if (create_flags & AUTHENTICATE_FLAG) {
+        os_ret = cmac_calc_start(ih->auth_ctx, key, _scratch_buf, scratch_buf_size);
+        if (os_ret) {
+            ret = MBED_ERROR_FAILED_OPERATION;
+            goto fail;
+        }
+        auth_started = true;
+        // Although key is not part of the data, we calculate CMAC on it as well
```
Comment (dannybenor) [inline features/storage/kvstore/securestore/SecureStore.cpp]: // Although name is not part of the data, we calculate CMAC on it as well

File: features/storage/kvstore/securestore/SecureStore.cpp [hunk 2]
```diff
@@ -0,0 +1,867 @@
...
+using namespace mbed;
+
+// --------------------------------------------------------- Definitions ----------------------------------------------------------
+
+static const uint32_t securestore_revision = 1;
+
+static const uint32_t enc_block_size    = 16;
+static const uint32_t cmac_size         = 16;
+static const uint32_t iv_size           = 8;
+static const uint32_t scratch_buf_size  = 32;
```
Comment (dannybenor) [inline features/storage/kvstore/securestore/SecureStore.cpp]: you can use a bigger buffer, maybe 256 bytes

File: features/storage/kvstore/securestore/SecureStore.cpp [hunk 3]
```diff
@@ -0,0 +1,867 @@
...
+
+    if (create_flags & AUTHENTICATE_FLAG) {
+        os_ret = cmac_calc_start(ih->auth_ctx, key, _scratch_buf, scratch_buf_size);
+        if (os_ret) {
+            ret = MBED_ERROR_FAILED_OPERATION;
+            goto end;
+        }
+        auth_started = true;
+
+        // Although key is not part of the data, we calculate CMAC on it as well
```
Comment (dannybenor) [inline features/storage/kvstore/securestore/SecureStore.cpp]: key-->name

File: features/storage/kvstore/securestore/mbed_lib.json [hunk 4]
```diff
@@ -0,0 +1,6 @@
+{
+    "name": "SecureStore",
+    "macros": ["MBEDTLS_CIPHER_MODE_CTR", "MBEDTLS_CMAC_C"],
+    "config": {
+    }
+}
```
Comment (dannybenor) [inline features/storage/kvstore/securestore/mbed_lib.json]: Something strange at the end of this file

**Review by dannybenor — CHANGES_REQUESTED** [review]
I reviewed securestore cpp and h
Fantastic Job!

File: features/storage/kvstore/securestore/mbed_lib.json (hunk 4 above)
Comment (davidsaada) [reply features/storage/kvstore/securestore/mbed_lib.json]: All I need here is to have the cipher and CMAC defines set. Nothing needed in the config.

File: features/storage/kvstore/securestore/SecureStore.cpp (hunk 2 above)
Comment (davidsaada) [reply features/storage/kvstore/securestore/SecureStore.cpp]: Good idea. Will check whether this has any effect on performance.

**Comment by dannybenor** [comment]
There is a strange character at the end of the file...

**Comment by davidsaada** [comment]
Fixed all issues. Enlarging the scratch buffer indeed helped with the performance, but that's just because there were less calls to TDBStore. Seems like TDBStore is the problem.

**Pull request closed without merging**